File: gaiag/baseline/Adaptor.js
// Dezyne --- Dezyne command line tools
//
// Copyright © 2015 Jan Nieuwenhuizen <janneke@gnu.org>
//
// This file is part of Dezyne.
//
// Dezyne is free software: you can redistribute it and/or modify it
// under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// Dezyne is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public
// License along with Dezyne.  If not, see <http://www.gnu.org/licenses/>.
//
// Commentary:
//
// Code:

dezyne.Adaptor = function(rt, meta) {
  this.rt = rt;
  rt.components = (rt.components || []).concat ([this]);
  this.meta = meta;
  this.flushes = true;
  this.State = {
    Idle: 0, Active: 1, Terminating: 2
  };
  this.State_to_string = {
    0: 'State_Idle', 1: 'State_Active', 2: 'State_Terminating'
  };
  this.state = this.State.Idle;
  this.count = 0;

  this.runner = new dezyne.IRun({provides: {name: 'runner', component: this}, requires: {}});
  this.choice = new dezyne.IChoice({provides: {}, requires: {name: 'choice', component: this}});

  this.runner.in.run = function() {
    runtime.call_in(this, function() {
      if(this.state === this.State.Idle && this.count < 2) {
        this.choice.in.e();
        this.state = this.State.Active;
      }
      else if(this.state === this.State.Idle && ! (this.count < 2)) { }
      else if(this.state === this.State.Active) {
        { }
      }
      else if(this.state === this.State.Terminating) { }
    }.bind(this), [this.runner, 'run']);
  }.bind(this);
  this.choice.out.a = function() {
    runtime.call_out(this, function() {
      if(this.state === this.State.Idle) { }
      else if(this.state === this.State.Active) {
        {
          this.count = this.count + 1;
          this.choice.in.e();
          this.state = this.State.Terminating;
        }
      }
      else if(this.state === this.State.Terminating && this.count < 2) {
        this.choice.in.e();
        this.state = this.State.Active;
      }
      else if(this.state === this.State.Terminating && ! (this.count < 2)) this.state = this.State.Idle;
    }.bind(this), [this.choice, 'a']);
  }.bind(this);

};

File: gaiag/baseline/Adaptor.cc
// Dezyne --- Dezyne command line tools
//
// Copyright © 2015 Jan Nieuwenhuizen <janneke@gnu.org>
//
// This file is part of Dezyne.
//
// Dezyne is free software: you can redistribute it and/or modify it
// under the terms of the GNU Affero General Public License as
// published by the Free Software Foundation, either version 3 of the
// License, or (at your option) any later version.
//
// Dezyne is distributed in the hope that it will be useful, but
// WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
// Affero General Public License for more details.
//
// You should have received a copy of the GNU Affero General Public
// License along with Dezyne.  If not, see <http://www.gnu.org/licenses/>.
//
// Commentary:
//
// Code:

#include "Adaptor.hh"

#include "locator.hh"
#include "runtime.hh"

#include <iostream>

namespace dezyne
{
  Adaptor::Adaptor(const locator& dezyne_locator)
  : dzn_meta{"","Adaptor",reinterpret_cast<const component*>(this),0,{},{[this]{runner.check_bindings();},[this]{console.check_bindings();}}}
  , dzn_rt(dezyne_locator.get<runtime>())
  , state(State::Idle)
  , count(0)
  , runner({{"runner",this},{"",0}})
  , console({{"",0},{"console",this}})
  {
    dzn_rt.performs_flush(this) = true; 
    runner.in.run = [&] () {
      call_in(this, [this] {runner_run();}, std::make_tuple(&runner, "run", "return"));
    };
    console.out.detected = [&] () {
      call_out(this, [this] {console_detected();}, std::make_tuple(&console, "detected", "return"));
    };
    console.out.deactivated = [&] () {
      call_out(this, [this] {console_deactivated();}, std::make_tuple(&console, "deactivated", "return"));
    };

  }

  void Adaptor::runner_run()
  {
    if (state == State::Idle and count < 2)
    {
      console.in.arm();
      state = State::Active;
    }
    else if (state == State::Idle and not (count < 2))
    {
    }
    else if (state == State::Active)
    {
      {
      }
    }
    else if (state == State::Terminating)
    {
      {
      }
    }
  }

  void Adaptor::console_detected()
  {
    if (state == State::Idle)
    {
      assert(false);
    }
    else if (state == State::Active)
    {
      {
        count = count + 1;
        console.in.disarm();
        state = State::Terminating;
      }
    }
    else if (state == State::Terminating)
    {
      assert(false);
    }
  }

  void Adaptor::console_deactivated()
  {
    if (state == State::Idle)
    {
      assert(false);
    }
    else if (state == State::Active)
    {
      assert(false);
    }
    else if (state == State::Terminating and count < 2)
    {
      console.in.arm();
      state = State::Active;
    }
    else if (state == State::Terminating and not (count < 2))
    state = State::Idle;
  }


  void Adaptor::check_bindings() const
  {
    dezyne::check_bindings(reinterpret_cast<const dezyne::component*>(this));
  }
  void Adaptor::dump_tree() const
  {
    dezyne::dump_tree(reinterpret_cast<const dezyne::component*>(this));
  }
}
